### DISCUSSION TITLE
vector<byte> support?

### DISCUSSION THREAD

**Discussion by jjqcat** [alibaba/yalantinglibs#676] [Q&A] [answered]
我想将一个结构体数据，通过byte数组传递，远程端手动再转换成我的结构体数据，有什么办法吗？是否可以通过vector<byte>进行转换，如
struct  custom
{
int a;
char c;
double d
};
vector<byte> vecData;
vecData.resize(sizeof(struct  custom));

struct  custom data;
memcpy(vecData.data, &data,sizeof(struct  custom));

RPC 函数调用时传递 vecData

谢谢

**Comment by poor-circle** [comment, marked as answer]
1.  struct_pack支持序列化你的结构体到`vector<byte>`
```cpp
custom data{};
auto buffer = struct_pack::serialize<std::vector<std::byte>>(data);
auto res = struct_pack::deserialize<std::vector<std::byte>>(buffer);
```
2. coro_rpc支持直接使用custom作为函数参数,不需要你手动序列化：
```cpp
void hi(custom data);
void call(coro_rpc::coro_rpc_client& client,custom data) {
     client.call<hi>(data);
}
```
3.  如果你特别想要手动接管序列化/反序列化，也不是不可以：
```cpp
void hi(std::span<std::byte> data/*or use std::vector<std::byte>&，with once more copy*/);
void call(coro_rpc::coro_rpc_client& client, std::vector<std::byte>& vec) {
     client.call<hi>(vec);
}
```
4. 也可以使用attachment传递数据：
 ```cpp
void hi() {
    auto ctx = coro_rpc::get_context();
    std::string_view attachment = ctx->get_request_attachment(); /*zero-copy by string_view*/
}
void call(coro_rpc::coro_rpc_client& client, std::string_view attachment) {
     client.set_req_attachment(attachment); /*zero-copy by string_view*/
     client.call<hi>();
}
```

**Comment by jjqcat** [comment]
我期望的是手动接管序列化/反序列化 返回值类型也可以是std::vector<std::byte> 吗

**Reply by poor-circle** [reply]
可以，不过建议使用set_resp_attachment/get_resp_attachement

**Reply by jjqcat** [reply]
好的，谢谢

**Comment by jjqcat** [comment]
验证 void call(coro_rpc::coro_rpc_client& client, std::vector<std::byte>& vec) ，链接通不过，编译可以，这个引用 & 去掉可以

**Reply by poor-circle** [reply]
链接不通过是不是因为你函数实现那里没加&
